packages/itwinui-react/src/core/DatePicker/DatePicker.tsx

```diff
@@ -699,11 +699,18 @@ export const DatePicker = React.forwardRef((props, forwardedRef) => {
                       role='option'
                       tabIndex={isSameDay(weekDay, focusedDay) ? 0 : -1}
                       aria-disabled={isDisabled ? 'true' : undefined}
-                      ref={(element) =>
-                        isSameDay(weekDay, focusedDay) &&
-                        needFocus.current &&
-                        element?.focus()
-                      }
+                      ref={(element) => {
+                        if (
+                          isSameDay(weekDay, focusedDay) &&
+                          needFocus.current
+                        ) {
+                          // Wait for DateRangeFilter's portaling to finish before focusing
+                          // See: https://github.com/iTwin/iTwinUI/pull/2297
+                          setTimeout(() => {
```

Comment:
Why not `queueMicrotask`?

Also I think the `if` check should be inside the microtask/timeout.

Reply:
> This is because I had noticed some failing tests with `queueMicrotask()`. I think this is because it conflicts with `FloatingFocusManager`'s focusing.

We might need a way to tell `FloatingFocusManager` to not do anything if we are already handling it.

Also, if only unit tests were failing, it could be because microtasks don't run reliably in unit tests. I had to add some hacky code when working on shadow DOM related tasks (see `setupTests`). You could try to convert these tests into e2e tests if it's giving you too much trouble.

> But actually after making these screen recordings, I noticed that the focus with `setTimeout()` starts from the trigger, goes to the previous month icon button, and then to the current date. I believe the focus should go from the trigger directly to the current date to prevent focus jumps and multiple callouts? Investigating further.

Yes, this was my concern with `setTimeout`. It might execute "too late" leading to focus jumping multiple times.

Reply:
> I went with the context approach to do this ([227827f](https://github.com/iTwin/iTwinUI/commit/227827f9668c1dd5727b3e265eec1d2a07faa593)). `Popover` adds a context containing the initialFocus around the floating content. The floating content can access this context and disable the initialFocus.

1. Whenever using context, make sure to memoize the `value` to prevent an unnecessary chain of re-renders.
2. Have you tested that this does not break the normal case where the first interactive element inside the popover is focused? (e.g. #2168)
3. (Maybe not relevant for this PR) We need to start thinking about a more general "overridable initial focus" solution that also works for Dialog.

Reply:
> Just wondering, if I directly put `setInitialFocus` instead of `{ setInitialFocus }` as the value in the context provider, would it still need to be memoized?

No, because `setInitialFocus` is already a stable reference. Whereas wrapping it in `{}` creates a new, unstable object during render (`{} === {}` is `false`).